File: problems/1288_removed_covered_interval.go
package main

import "sort"

type byRange [][]int

func (a byRange) Len() int {
	return len(a)
}
func (a byRange) Swap(i, j int) {
	a[i], a[j] = a[j], a[i]
}
func (a byRange) Less(i, j int) bool {
	return a[i][0] < a[j][0] || (a[i][0] == a[j][0] && a[i][1] > a[j][1])
}

func removeCoveredIntervals(intervals [][]int) int {
	sort.Sort(byRange(intervals))
	n := len(intervals)
	total := n
	current := intervals[0]
	for i := 1; i < n; i++ {
		if current[1] >= intervals[i][1] {
			total--
		} else {
			current = intervals[i]
		}
	}
	return total
}

/*
Custom Sort in Go

package main

import (
    "fmt"
    "sort"
)

type byLength []string

func (s byLength) Len() int {
    return len(s)
}
func (s byLength) Swap(i, j int) {
    s[i], s[j] = s[j], s[i]
}
func (s byLength) Less(i, j int) bool {
    return len(s[i]) < len(s[j])
}

func main() {
    fruits := []string{"peach", "banana", "kiwi"}
    sort.Sort(byLength(fruits))
    fmt.Println(fruits)
}
*/

File: problems/1288_removed_covered_interval.cpp
#include<vector>
#include<iostream>
using namespace std;
class Solution {
public:
    int removeCoveredIntervalsOrig(vector<vector<int>>& intervals) {
        sort(intervals.begin( ),
             intervals.end( ),
             [ ](const vector<int>& lhs, const vector<int>& rhs)
             { return lhs[0] < rhs[0] || (lhs[0] == rhs[0] && lhs[1] > rhs[1]); });
        int num_removed = 0;
        for(int i = 0; i < intervals.size(); i++){
            // cout << "[" << intervals[i][0] << "," << intervals[i][1] << "]  ";
            if (i > 0 && (intervals[i][1] <= intervals[i-1][1] ||
                          intervals[i][0] == intervals[i-1][0])) {
                intervals[i-1][1] = std::max(intervals[i-1][1], intervals[i][1]);
                num_removed++;
            }
        }
        return intervals.size() - num_removed;

    }
    int removeCoveredIntervals(vector<vector<int>>& intervals) {
        sort(intervals.begin( ),
             intervals.end( ),
             [&](const vector<int>& lhs, const vector<int>& rhs)
             { return lhs[0] < rhs[0] || (lhs[0] == rhs[0] && lhs[1] > rhs[1]); });
        int total = intervals.size();
        vector<int> current = intervals[0];
        for(int i = 1; i < intervals.size(); i++){
            if (current[1] >= intervals[i][1]){
                total--;
            } else {
                current = intervals[i];
            }
        }
        return total;
    }

    /*
    Sort w/Lambda C++ Example:

sort(intervals.begin( ),
     intervals.end( ),
     [ ]( const vector<int>& lhs, const vector<int>& rhs )
    {
       return lhs[0] < rhs[0];
    }
);
    */
};


/*

[[1,4],[3,6],[2,8]]


Good test case:
[[66672,75156],[59890,65654],[92950,95965],[9103,31953],[54869,69855],[33272,92693],[52631,65356],[43332,89722],[4218,57729],[20993,92876]]

*/
